File: tree/653.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findTarget(root *TreeNode, k int) bool {
    l := make([]int, 0)
    rootl := make([]*TreeNode, 0)
    for {
        for root != nil {
            rootl = append(rootl, root)
            root = root.Left
        }
        if len(rootl) == 0 {
            break
        }
        root = rootl[len(rootl)-1]
        l = append(l, root.Val)
        rootl = rootl[:len(rootl)-1]
        root = root.Right
    }

    i := 0
    j := len(l) - 1
    for i < j {
        if l[i] + l[j] == k {
            return true
        }else if l[i] + l[j] < k {
            i++
        }else {
            j--
        }
    }
    return false
    
}

File: tree/653.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        std::stack<TreeNode*> s;
        s.emplace(root);
        std::unordered_map<int, int> m;
        while (!s.empty()) {
            root = s.top();
            s.pop();
            auto iter = m.find(k-root->val);
            if (iter != m.end()) {
                return true;
            } else {
                m[root->val] = 1;
            }
            if (root->right)
                s.emplace(root->right);
            if (root->left)
                s.emplace(root->left);
        }
        return false;
    }
};
